File: source/render/triangle.cpp
#include "render/triangle.h"


kl::triangle::triangle()
{}

kl::triangle::triangle(const vertex& a, const vertex& b, const vertex& c) : a(a), b(b), c(c)
{}

kl::float4 kl::triangle::get_constants() const
{
    const float constants = 1.0f / ((b.world.y - c.world.y) * (a.world.x - c.world.x) + (c.world.x - b.world.x) * (a.
        world.y - c.world.y));
    return {
        (b.world.y - c.world.y) * constants,
        (c.world.x - b.world.x) * constants,
        (c.world.y - a.world.y) * constants,
        (a.world.x - c.world.x) * constants
    };
}

kl::float3 kl::triangle::get_weights(const float4& constants, const float2& point) const
{
    const float dx = point.x - c.world.x;
    const float dy = point.y - c.world.y;
    const float w1 = dx * constants.x + dy * constants.y;
    const float w2 = dx * constants.z + dy * constants.w;
    return {w1, w2, 1.0f - w1 - w2};
}

kl::float3 kl::triangle::get_weights(const float3& position) const
{
    const float3 v0 = a.world - c.world;
    const float3 v1 = b.world - c.world;
    const float3 v2 = position - c.world;
    
    const float d00 = v0.dot(v0);
    const float d01 = v0.dot(v1);
    const float d11 = v1.dot(v1);
    const float d20 = v2.dot(v0);
    const float d21 = v2.dot(v1);
    
    const float inverse_denom = 1.0f / (d00 * d11 - d01 * d01);
    const float w1 = (d11 * d20 - d01 * d21) * inverse_denom;
    const float w2 = (d00 * d21 - d01 * d20) * inverse_denom;
    
    return {w1, w2, 1.0f - w1 - w2};
}

bool kl::triangle::is_in_triangle(const float3& weights)
{
    return !(weights.x < 0.0f || weights.y < 0.0f || weights.z < 0.0f);
}

float kl::triangle::interpolate(const float3& weights, const float3& values)
{
    return weights.dot(values);
}

kl::vertex kl::triangle::interpolate(const float3& weights) const
{
    return {
        float3{
            interpolate(weights, {a.world.x, b.world.x, c.world.x}),
            interpolate(weights, {a.world.y, b.world.y, c.world.y}),
            interpolate(weights, {a.world.z, b.world.z, c.world.z})
        },
        float2{
            interpolate(weights, {a.texture.x, b.texture.x, c.texture.x}),
            interpolate(weights, {a.texture.y, b.texture.y, c.texture.y})
        },
        float3{
            interpolate(weights, {a.normal.x, b.normal.x, c.normal.x}),
            interpolate(weights, {a.normal.y, b.normal.y, c.normal.y}),
            interpolate(weights, {a.normal.z, b.normal.z, c.normal.z})
        }
    };
}


std::ostream& kl::operator<<(std::ostream& os, const triangle& obj)
{
    os << "{" << obj.a << ", " << obj.b << ", " << obj.c << "}";
    return os;
}

File: source/kl/render/Triangle.java
package kl.render;

import kl.math.*;

public class Triangle {
    public Vertex a, b, c;

    public Triangle() {
        a = new Vertex();
        b = new Vertex();
        c = new Vertex();
    }
    public Triangle(Vertex a, Vertex b, Vertex c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }
    public Triangle(Triangle t) {
        a = new Vertex(t.a);
        b = new Vertex(t.b);
        c = new Vertex(t.c);
    }

    public Float4 getConstants() {
	    final float tempConst = 1.0f / ((b.world.y - c.world.y) * (a.world.x - c.world.x) + (c.world.x - b.world.x) * (a.world.y - c.world.y));
        return new Float4(
            (b.world.y - c.world.y) * tempConst,
            (c.world.x - b.world.x) * tempConst,
            (c.world.y - a.world.y) * tempConst,
            (a.world.x - c.world.x) * tempConst
        );
    }
    public Float3 getWeights(Float4 constants, Float2 position) {
        final float dx = position.x - c.world.x;
        final float dy = position.y - c.world.y;
        final float w1 = dx * constants.x + dy * constants.y;
        final float w2 = dx * constants.z + dy * constants.w;
        return new Float3(w1, w2, 1.0f - w1 - w2);
    }
    public Float3 getWeights(Float3 point) {
        final Float3 v0 = new Float3(a.world.subtract(c.world));
        final Float3 v1 = new Float3(b.world.subtract(c.world));
        final Float3 v2 = new Float3(point.subtract(c.world));
        final float d00 = v0.dot(v0);
        final float d01 = v0.dot(v1);
        final float d11 = v1.dot(v1);
        final float d20 = v2.dot(v0);
        final float d21 = v2.dot(v1);
        final float invDenom = 1.0f / (d00 * d11 - d01 * d01);
        final float w1 = (d11 * d20 - d01 * d21) * invDenom;
        final float w2 = (d00 * d21 - d01 * d20) * invDenom;
        return new Float3(w1, w2, 1.0f - w1 - w2);
    }

    public boolean in(Float3 weights) {
        return !(weights.x < 0.0f || weights.y < 0.0f || weights.z < 0.0f);
    }

    public float interpolate(Float3 weights, Float3 values) {
        return weights.dot(values);
    }
    public Vertex interpolate(Float3 weights) {
        return new Vertex(
            new Float3(
                interpolate(weights, new Float3(a.world.x, b.world.x, c.world.x)),
                interpolate(weights, new Float3(a.world.y, b.world.y, c.world.y)),
                interpolate(weights, new Float3(a.world.z, b.world.z, c.world.z))
            ),
            new Float2(
                interpolate(weights, new Float3(a.texture.x, b.texture.x, c.texture.x)),
                interpolate(weights, new Float3(a.texture.y, b.texture.y, c.texture.y))
            ),
            new Float3(
                    interpolate(weights, new Float3(a.normal.x, b.normal.x, c.normal.x)),
                    interpolate(weights, new Float3(a.normal.y, b.normal.y, c.normal.y)),
                    interpolate(weights, new Float3(a.normal.z, b.normal.z, c.normal.z))
            )
        );
    }

    public String toString() {
        return "{" + a + ", " + b + ", " + c + "}";
    }
}
